### DISCUSSION TITLE
Animating a continous loop, or rather a set of points, point for point

### DISCUSSION THREAD

**Discussion by HafeezHaidari** [marcomusy/vedo#866] [Q&A] [answered]
Hi guys, ive recently started a project in python but im already stuck.
I wanted to create an app that animates the lorenz-attractor in real time. I looked into a couple of libraries and decided for vedo. But now I’m stuck in the animating part. How do i simply cause the points (which are in a multidimensional numpy array) to be drawn one by one and animated.
Here is the code i already have:
```py
from vedo import *


def createPoints(n):
    pts = np.empty((1, 3))
    a = 10
    b = 28
    c = 8 / 3
    x = 0.01
    y = 0
    z = 0
    iterations = 0
    while iterations < n:
        dt = 0.01
        dx = (a * (y - x)) * dt
        dy = (x * (b - z)) * dt
        dz = (x * y - c * z) * dt
        x = x + dx
        y = y + dy
        z = z + dz
        pts = np.append(pts, np.array([[x, y, z]]), axis=0)
        iterations += 1
    return pts


points = Points(createPoints(1000), r=2)
show(points, __doc__, axes=1)
```

**Comment by marcomusy** [comment]
> I looked into a couple of libraries and decided for vedo

Excellent choice LOL

You can do something like:
```py

from vedo import *

def create_points(n):
    pts = np.empty((1, 3))
    a = 10
    b = 28
    c = 8 / 3
    x = 0.01
    y = 0
    z = 0
    iterations = 0
    while iterations < n:
        dt = 0.01
        dx = (a * (y - x)) * dt
        dy = (x * (b - z)) * dt
        dz = (x * y - c * z) * dt
        x = x + dx
        y = y + dy
        z = z + dz
        pts = np.append(pts, np.array([[x, y, z]]), axis=0)
        iterations += 1
    return pts

pts = create_points(1000)
line = Line(pts)
pt = Point(pts[0], c='red5', r=15)
comment = Text2D(font="Calco")

plt = Plotter(axes=1, interactive=False)
plt.show(line, pt, comment, viewup='z')

# Solution1 ######################
# Animate the point along the curve 
# (but cannot interact with the plot)
# for i, p in progressbar(enumerate(pts)):
#     pt.pos(p)
#     comment.text(f"Iteration: {i}")
#     plt.render()

# Solution2 ######################
# can interact with the plot
def loop_func(event):
    global i
    i += 1
    if i >= len(pts):
        plt.timer_callback('stop', tid)
        return
    pt.pos(pts[i])
    comment.text(f"Iteration: {i}")
    plt.render()
i = 0       
plt.add_callback("timer", loop_func)
tid = plt.timer_callback("start")

# stay interactive until user presses q or close the window
plt.interactive()
```
![Screenshot from 2023-05-23 18-58-30](https://github.com/marcomusy/vedo/assets/32848391/ad668e25-42a7-4c8c-9987-0b3ce4522291)

**Reply by HafeezHaidari** [reply]
I actually like solution 2 alot, but im trying to sort of draw the points one by one or a moving line rather than having a line and a dot following it. 
Something like this: https://media.geeksforgeeks.org/wp-content/uploads/Lorenz_Attractor_Output.gif
Additionally: Are there any resources for me to learn vedo with?

**Reply by marcomusy** [reply, marked as answer]
`vedo` has a built in feature to draw trailing lines, (you can set `n=1000` to get the full trajectory):
```py 
from vedo import *

def create_points(n):
    a = 10
    b = 28
    c = 8 / 3
    dt = 0.01
    p = np.array([0.01, 0, 0])
    pts = []
    for _ in range(n):
        x, y, z = p
        dx = (a * (y - x)) * dt
        dy = (x * (b - z)) * dt
        dz = (x * y - c * z) * dt
        p = p + [dx, dy, dz]
        pts.append(p)
    return np.array(pts)

pts = create_points(1000)
line = Line(pts).c("k5")
pt = Point(pts[0], c="red5", r=15)
pt.add_trail(c="red3", lw=4, n=200)
comment = Text2D(font="Calco")

plt = Plotter(axes=1, interactive=False)
plt.show(line, pt, comment, viewup="z", zoom=1.2)

def loop_func(event):
    global i
    i += 1
    if i >= len(pts):
        plt.timer_callback("stop", tid)
        return
    pt.pos(pts[i])._update_trail()
    comment.text(f"Iteration: {i}")
    plt.render()

i = 0
plt.add_callback("timer", loop_func)
tid = plt.timer_callback("start")
plt.interactive()
```
<img width="365" alt="Screenshot 2023-05-24 at 15 59 04" src="https://github.com/marcomusy/vedo/assets/32848391/5c5222eb-b65a-4ac7-a04f-33c3f93a58f0">

  

> Additionally: Are there any resources for me to learn vedo with?

- This is a very general introduction/overview of the library (30 minutes video):
https://www.youtube.com/watch?v=MhIoetdxwc0&t=36s&pp=ygULcHl0aG9uIHZlZG8%3D

- One effective way to learn `vedo` is to look for any [examples](https://vedo.embl.es/#gallery) that look interesting to you in the "basic" directory and run them from your command line with:
```
vedo --run name_of_example
```
you will get a printout of the code that generates the output. These examples are thought of to be sort of "building blocks" of what you can do with the library. The API is documented [here](https://vedo.embl.es/docs).

- Ask the maintainers :)
